src/realm/object-store/shared_realm.cpp

```diff
@@ -1278,11 +1313,15 @@ void Realm::close()
     do_invalidate();
 
     m_binding_context = nullptr;
+    if (policy == Realm::ClosePolicy::ManualReleaseRealmHandle) {
```

Comment:
Cope with failures that might happen within a callback. Most of the time it is not needed to explicitly release resources via the coordinator (when the object goes out of scope, eventually the file handle will be released). However, in case of exceptions originated within initialization callback, we do need to  to release the handle immediately in order to delete Realm files.

Reply:
This is not really a valid thing to do. Coordinators are shared between multiple threads, and calling `RealmCoordinator::close()` while it's being used by other threads will crash. If it's not currently being used by anything else, explicitly closing it doesn't do anything that releasing the reference to it isn't doing already.

Reply:
A perfectly reasonable thing to do might be something like the following:

```
std::vector<std::thread> threads;
for (int i = 0; i < 10; ++i) {
    threads.emplace_back([&] {
        auto realm = Realm::get_shared_realm(config);
        // do some processing
    });
}
```

If this is the first open and config has an initialization function, we guarantee that the initialization function will be called exactly once and none of the Realm instances will be returned from `get_shared_realm()` until after it's run. This works in multi-process scenarios as well.

This is achieved using the Realm's write lock. One of the threads "wins" and acquires the write lock and initializes the schema and then calls the initialization function and commits that. Each of the other threads then acquire the write lock, see that the schema has been initialized by someone else, and then release it without doing anything. Closing the DB while they're waiting for the write lock means that they'll instead break in awful ways.

It's not obvious to me that the requested functionality of deleting the Realm on initialization failure is actually possible to do safely. #5364 is asking for that as a workaround for something that I'd expect to already work. If the initialization function throws an exception we don't commit the transaction which initialized the schema, so the next open should call the initialization function again even though the Realm file already exists.

Reply:
@nirinchev @cmelchior  is this the case (the broken transaction is committed)? I think @tgoyne  is right. The code in this PR will only work if there is only one thread and crash the other threads if there are multiple threads that are trying to do "get_shared_realm"